### ISSUE TITLE
An autograd hack?

### ISSUE THREAD

**Issue by jatentaki** [stevenygd/PointFlow#27]
I'm trying to figure out the role of [this line](https://github.com/stevenygd/PointFlow/blob/master/models/networks.py#L156). It looks like some autograd hack similar to straight through estimators of form

```python
logits = ...
smooth_probs = torch.sigmoid(logits)
hard_samples = (logits > 0).to(logits.dtype)

# values come from `hard_samples` but gradient from `smooth_probs`
binary_straight_through = hard_samples + smooth_probs - smooth_probs.detach()
```

 but in this case it really seems to be a no-op. Could I ask for some explanation?

Thank you

**Comment by stevenygd** [comment]
Yes this is an autograd hack to make the computational graph sequential. We need this because at that time (pytorch 1.4 I believed) there was concurrency bug in pytorch - training for long enough can see a race condition on a C++ variable that stores information about enable/disable gradient flow.
